src/Analysis/Engine/Impl/AnalysisHashSet.cs

```diff
@@ -275,7 +275,19 @@ public AnalysisHashSet(IEnumerable<AnalysisValue> enumerable, IEqualityComparer<
                 return true;
             }
 
+            // Only access Count once to prevent wasted time in bucket locks.
+            var lCount = Count;
+            var rCount = other.Count;
+
+            if (lCount == 0 && rCount == 0) {
+                return true;
+            }
+
             if (Comparer == other.Comparer) {
+                if (lCount != rCount) {
```

Comment:
This optimization may not work. Code below gets hash codes and collapses them into hash sets, which means that two buckets arrays with different amount of items will be considered equal. I can't say if that is correct or not.

Reply:
If the comparers are the same, should there ever be a situation where two sets have different contents that are the same? To me that seems related to the issue of keys sometimes changing after being added (that `Debug.Fail`), except that this specific check is costly to do.

Reply:
> Or does a union not actually do any unioning and instead concatenates?

I've seen sets with empty buckets under debugger, but I have no idea how to reproduce it and I don't know if that is a buggy result or an expected one. This code is quite muddy. @MikhailArkhipov , what is your opinion?

> Namely, I started with the count checks, then saw a lot of CPU time used on accessing Count (since it's put behind a lock), then moved the access above.

Just for the reference. [Thin locks](https://blogs.msdn.microsoft.com/seteplia/2017/09/06/managed-object-internals-part-2-object-header-layout-and-the-cost-of-locking/) are really fast, but I don't know if we have thin or fat lock here. `GetHashCode` is overridden, but it isn't the only thing that can promote thin lock to fat lock.

> The MRO lookup change curbs things in some cases.

Yes, MRO fix is clear.

Reply:
The locking I'm referring to is here: https://github.com/Microsoft/python-language-server/blob/cb1c3dfbad1c7a521857a5dae8eea7c94225a096/src/Analysis/Engine/Impl/AnalysisHashSet.cs#L95-L106

Now that I look more closely, most other functions grab `Bucket` at the start. The original function here did (by grabbing `lBuckets` and maybe grabbing `rBuckets`). Perhaps I should have moved those upward and operated on them instead of introducing two extra possible locks.

I'm happy to remove some of the more suspect length checks if you think this may be an issue (but some of them should still be valid).

Reply:
Oh, I see, the thin vs fat locking is past `lock` and more to do with object layout, so me showing where the locking is doesn't matter. Oops.

Reply:
Different numbers of items (types) in buckets and yet equality seem odd to me. I am not overly concerned with keeping correct number of items in buckets since collecting types from various type of calls is generally not what we want to be doing. Ex, `def f(a): return a` called with `int` now has int return. Called then with `str` it is suddenly show as returning `int, str`. But that's not true, it is not returning combo or a union. And then user adds code calling with bool and it now it returns triplet... Which has nothing to do with the nature of the code.

Reply:
New analysis has special types for call arguments and generics so the said function is considered returning type of its argument so no need to collect possible outcomes. `f(1).` will show completion for `int` members and `f('a').` will yield `str` members without need to hold unions.

Reply:
In general LGTM. I'd build insiders and let people try it out.